File: 프로그래머스/LV2/코딩테스트 연습/2018 KAKAO BLIND RECRUITMENT/[1차] 뉴스 클러스터링/[1차] 뉴스 클러스터링.py
def solution(str1, str2):
    str1 = str1.upper()
    str2 = str2.upper()
    A = [str1[i:i+2] for i in range(len(str1)-1) if str1[i].isalpha() and str1[i+1].isalpha()]
    B = [str2[i:i+2] for i in range(len(str2)-1) if str2[i].isalpha() and str2[i+1].isalpha()]
    AS = set(A)
    BS = set(B)
    intersection = AS & BS
    union = AS | BS
    numerator = sum([min(A.count(ab), B.count(ab)) for ab in intersection])
    denominator = sum([max(A.count(ab), B.count(ab)) for ab in union])

    JAB = 1 if denominator == 0 and numerator == 0 else numerator / denominator
    return int(JAB * 65536)

File: 프로그래머스/LV2/코딩테스트 연습/2018 KAKAO BLIND RECRUITMENT/[1차] 뉴스 클러스터링/[1차] 뉴스 클러스터링.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <set>
#include <cctype>
#include <iterator>
using namespace std;

int solution(string str1, string str2)
{
    int numerator = 0;
    int denominator = 0;
    double JAB;
    vector<string> A, B, intersection, uni;

    transform(str1.begin(), str1.end(), str1.begin(), ::toupper);
    transform(str2.begin(), str2.end(), str2.begin(), ::toupper);
    for (int i = 0; i <= str1.length() - 2; i++)if(isalpha(str1[i]) && isalpha(str1[i+1]))A.push_back(str1.substr(i, 2));
    for (int i = 0; i <= str2.length() - 2; i++)if(isalpha(str2[i]) && isalpha(str2[i+1]))B.push_back(str2.substr(i, 2));
    
    set<string> AS(A.begin(), A.end());
    set<string> BS(B.begin(), B.end());

    set_intersection(AS.begin(), AS.end(),BS.begin(), BS.end(),back_inserter(intersection));
    set_union(AS.begin(), AS.end(),BS.begin(), BS.end(),back_inserter(uni));

    for (string ab : intersection)numerator += min(count(A.begin(), A.end(), ab), count(B.begin(), B.end(), ab));
    for (string ab : uni)denominator += max(count(A.begin(), A.end(), ab), count(B.begin(), B.end(), ab));

    JAB = numerator == 0 && denominator == 0 ? 1 : (double)numerator / (double)denominator;
    return (int)(JAB * 65536);
}